pageserver/src/http/routes.rs

```diff
@@ -1423,6 +1425,76 @@ async fn list_location_config_handler(
     json_response(StatusCode::OK, result)
 }
 
+// Do a time travel recovery on the given tenant/tenant shard. Tenant needs to be detached
+// (from all pageservers) as it invalidates consistency assumptions.
+async fn tenant_time_travel_remote_storage_handler(
+    request: Request<Body>,
+    cancel: CancellationToken,
+) -> Result<Response<Body>, ApiError> {
+    let tenant_shard_id: TenantShardId = parse_request_param(&request, "tenant_shard_id")?;
+
+    check_permission(&request, Some(tenant_shard_id.tenant_id))?;
+
+    let timestamp_raw = must_get_query_param(&request, "travel_to")?;
+    let timestamp = humantime::parse_rfc3339(&timestamp_raw)
+        .with_context(|| format!("Invalid time for travel_to: {timestamp_raw:?}"))
+        .map_err(ApiError::BadRequest)?;
+
+    let done_if_after_raw = must_get_query_param(&request, "done_if_after")?;
+    let done_if_after = humantime::parse_rfc3339(&done_if_after_raw)
+        .with_context(|| format!("Invalid time for done_if_after: {done_if_after_raw:?}"))
+        .map_err(ApiError::BadRequest)?;
+
+    // This is just a sanity check to fend off naive wrong usages of the API:
+    // the tenant needs to be detached *everywhere*
+    let state = get_state(&request);
+    let we_manage_tenant = state.tenant_manager.manages_tenant_shard(tenant_shard_id);
+    if we_manage_tenant {
+        return Err(ApiError::BadRequest(anyhow!(
+            "Tenant {tenant_shard_id} is already attached at this pageserver"
+        )));
+    }
+
+    let Some(storage) = state.remote_storage.as_ref() else {
+        return Err(ApiError::InternalServerError(anyhow::anyhow!(
+            "remote storage not configured, cannot run time travel"
+        )));
+    };
+
+    if timestamp > done_if_after {
+        return Err(ApiError::BadRequest(anyhow!(
+            "The done_if_after timestamp comes before the timestamp to recover to"
+        )));
+    }
+
+    tracing::info!("Issuing time travel request internally. timestamp={timestamp_raw}, done_if_after={timestamp_raw}");
```

Comment:
```suggestion
    tracing::info!("Issuing time travel request internally. timestamp={timestamp}, done_if_after={done_if_after}");
```

Reply:
```
error[E0277]: `SystemTime` doesn't implement `std::fmt::Display`
    --> pageserver/src/http/routes.rs:1470:71
     |
1470 |     tracing::info!("Issuing time travel request internally. timestamp={timestamp}, done_if_after={timestamp_raw}");
     |                                                                       ^^^^^^^^^^^ `SystemTime` cannot be formatted with the default formatter
     |
     = help: the trait `std::fmt::Display` is not implemented for `SystemTime`
     = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
     = note: this error originates in the macro `format_args` which comes from the expansion of the macro `tracing::info` (in Nightly builds, run with -Z macro-backtrace for more info)
```

And the debug impl is [quite undesireable](https://github.com/rust-lang/rust/blob/11f32b73e0dc9287e305b5b9980d24aecdc8c17f/library/std/src/sys/pal/unix/time.rs#L64-L71). One could of course use the humantime crate to format the data, but why do that when it's already present in a readable format?

I mainly added that log line so that the parameters sent to the request are known. Adding them to the span seemed like overkill to me as that would put them to every single log line.

Reply:
On pageserver side we do have the chrono as well. Aah yes, please note the same timestamp was on the line twice. I'm fine logging the string version.

Reply:
Made it print `done_if_after_raw`